### PR TITLE
Add gitlab utils 1

### PR DESCRIPTION
**Pull request by schettn**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is no method to get user specific data from GitLab#7

**What is the new behavior (if this is a feature change)?**
- Now there is a method to get user specific data from GitLab.

### CODE REVIEW

File: src/utilities/GitLabUtils/index.js [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  const fillStructure = async (base) => {
+    // Fill profile of a user
+    const fillProfile =  async () => {
+      const url = `https://${base.platformUrl}/${username}`;
+      parseTextToDOM(fetchHtml(url)).then(html => {
+        //console.log(html)
+        const status = html.getElementsByTagName("gl-emoji")[0];
+        const coverDesc =  html.getElementsByClassName("cover-desc")[0].getElementsByTagName("span");
+        const coverTitle =  html.getElementsByClassName("cover-title")[0];
+        const avatarUrl = html.getElementsByClassName("avatar-holder")[0].getElementsByTagName('a')[0].getAttribute('href');
```
Comment (Aichnerc) [inline src/utilities/GitLabUtils/index.js]: Please follow our coding guidelines on max. char-length per line policy.

File: src/utilities/GitLabUtils/index.js [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+};
+
+
+//> Member functions
+// Get a member object by username
+const getMember = async (username) => {
+  const url = `https://${data.platformUrl}/${username}`
+  return await parseTextToDOM(fetchHtml(url)).then(html => {
+    let member = {...members}
+    const avatarUrl = html.getElementsByClassName("avatar-holder")[0].getElementsByTagName('a')[0].getAttribute('href');
```
Comment (Aichnerc) [inline src/utilities/GitLabUtils/index.js]: Please follow our coding guidelines on max. char-length per line policy.

File: src/utilities/GitLabUtils/index.js [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+      }
+
+      //> Contributions
+      // Convert HTML formatted event-items to contributions
+      const convertToContributions = async (items) => {
+        let contributions = {}
+        for(const element of items){
+          let contrib = {...contribution};
+          var time = element.getElementsByTagName('time')[0].getAttribute("datetime");
+          var nameWithOwner = element.getElementsByClassName('event-scope')[0].getElementsByTagName('a')[0].getAttribute("href");
```
Comment (Aichnerc) [inline src/utilities/GitLabUtils/index.js]: Please follow our coding guidelines on max. char-length per line policy.

File: src/utilities/GitLabUtils/index.js [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+  .then((res) => {
+    return parser.parseFromString(res.html, "text/html");
+  });
+  return html;
+}
+//Parse plain text to DOM Object
+const parseTextToDOM = (json) => {
+  const parser = new DOMParser();
+  const html = json
+  .then((res) => {
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: Please put then in the same line as the object.
No brackets needed.

```javascript
const parseTextToDOM = json => {
  const parser = new DOMParser();
  const html = json.then(res => {
    return parser.parseFromString(res, "text/html");
  });
  return html;
};

```

File: src/utilities/GitLabUtils/index.js [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+  })
+  .then( (res) =>  res.text())
+  .then( (res) => {return res;});
+}
+
+//> Member functions
+// Get a member object by username
+const getMember = async (username) => {
+  const url = `https://${data.platformUrl}/${username}`;
+  return await parseTextToDOM(fetchHtml(url)).then((html) => {
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: No brackets needed.

```javascript
const getMember = async username => {
  const url = `https://${data.platformUrl}/${username}`;
  return await parseTextToDOM(fetchHtml(url)).then(html => {
    let member = { ...members };
    const avatarUrl = html
      .getElementsByClassName("avatar-holder")[0]
      .getElementsByTagName("a")[0]
      .getAttribute("href");
    const name = html.getElementsByClassName("cover-title")[0].innerHTML;

    member.name = name;
    member.username = username;
    member.avatarUrl = `https://${data.platformUrl}/${avatarUrl.substring(1)}`;
    member.webUrl = `https://${data.platformUrl}/${username}`;

    return member;
  });
};
```

File: src/utilities/GitLabUtils/index.js [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+      //"X-Requested-With": "XMLHttpRequest"
+    }
+  })
+  .then( (res) =>  res.text())
+  .then( (res) => {return res;});
+}
+
+//> Member functions
+// Get a member object by username
+const getMember = async (username) => {
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: No brackets needed.

```javascript
const getMember = async username => {
  const url = `https://${data.platformUrl}/${username}`;
  return await parseTextToDOM(fetchHtml(url)).then(html => {
    let member = { ...members };
    const avatarUrl = html
      .getElementsByClassName("avatar-holder")[0]
      .getElementsByTagName("a")[0]
      .getAttribute("href");
    const name = html.getElementsByClassName("cover-title")[0].innerHTML;

    member.name = name;
    member.username = username;
    member.avatarUrl = `https://${data.platformUrl}/${avatarUrl.substring(1)}`;
    member.webUrl = `https://${data.platformUrl}/${username}`;

    return member;
  });
};
```

File: src/utilities/GitLabUtils/index.js [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+// Get Json data file
+export const get = async (server, username) => {
+
+  const fillStructure = async (base) => {
+    // Fill profile of a user
+    const fillProfile =  async () => {
+      const url = `https://${base.platformUrl}/${username}`;
+      parseTextToDOM(fetchHtml(url)).then((html) => {
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: No brackets needed.

```javascript
// Get Json data file
export const get = async (server, username) => {
  const fillStructure = async base => {
    // Fill profile of a user
    const fillProfile = async () => {
      const url = `https://${base.platformUrl}/${username}`;
      parseTextToDOM(fetchHtml(url)).then(html => {
        //console.log(html)
        const status = html.getElementsByTagName("gl-emoji")[0];
        const coverDesc = html
          .getElementsByClassName("cover-desc")[0]
          .getElementsByTagName("span");
        const coverTitle = html.getElementsByClassName("cover-title")[0];
        const avatarUrl = html
          .getElementsByClassName("avatar-holder")[0]
          .getElementsByTagName("a")[0]
          .getAttribute("href");
        const links = html.getElementsByClassName("profile-link-holder");
        const message = null;
        const emojiHTML = null;
        const username = coverDesc[0].innerHTML.trim().substring(1);
        const date = coverDesc[1].innerHTML;

        base.avatarUrl = `https://${data.platformUrl}/${avatarUrl.substring(
          1
        )}`;

        base.username = username;
        base.name = coverTitle.innerHTML.trim();
        base.createdAt = new Date(date);
      });
    };
```

File: src/utilities/GitLabUtils/index.js [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+          base.status.emojiHTML = status.outerHTML;
+        }
+        */
+
+        base.avatarUrl = `https://${data.platformUrl}/${avatarUrl.substring(1)}`;
+
+        base.username = username;
+        base.name = coverTitle.innerHTML.trim();
+        base.createdAt = new Date(date);
+
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: No newline needed, please remove.

File: src/utilities/GitLabUtils/index.js [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+      });
+    }
+    // Fill all Organizations a user is in into structure
+    const fillOrganizations =  async () => {
+      const url = `https://${base.platformUrl}/users/${username}/groups.json`;
+      //console.log(url)
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/index.js]: Please remove all console.log comments.

File: src/utilities/GitLabUtils/Objects/index.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
+export const structure = {
+    "platformName": "GitLab",
```
Comment (kleberbaum) [inline src/utilities/GitLabUtils/Objects/index.js]: No need for double quotes with keys.

```javascript
export const structure = {
  platformName: "GitLab",
  platformUrl: "https://gitlab.com",
  avatarUrl: "",
  websiteUrl: "",
  company: "",
  email: "",
  name: "",
  createdAt: "",
  location: "",
  status: {
    message: "",
    emojiHTML: ""
  },
  organizations: [],
  repos: [],
  contributions: {
    years: []
  }
};
```

File: src/utilities/GitLabUtils/index.js (hunk 9 above)
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Or mark them as important debugging points with a comment heading and description.

File: src/utilities/GitLabUtils/index.js (hunk 7 above)
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Actually it is a good practice and code quality thing to add brackets.

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/utilities/GitLabUtils/index.js (hunk 6 above)
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Actually it is a good practice and code quality thing to add brackets.

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/utilities/GitLabUtils/index.js (hunk 5 above)
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Actually it is a good practice and code quality thing to add brackets.

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/utilities/GitLabUtils/index.js [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+//> Parser functions
+// Parse Json to DOM Object
+const parseJsonToDOM = (json) => {
+  const parser = new DOMParser();
+  const html = json
+  .then((res) => {
+    return parser.parseFromString(res.html, "text/html");
+  });
+  return html;
+}
```
[the remark this answers is not in the document]
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: There is also a missing bracket around `json`

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/utilities/GitLabUtils/index.js [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+// Import JSON structure
+import { structure, organizations, members, contribution, repos, calendarEntry, yearEntry} from "./Objects";
+let data = structure;
+
+//> Parser functions
+// Parse Json to DOM Object
+const parseJsonToDOM = (json) => {
+  const parser = new DOMParser();
+  const html = json
+  .then((res) => {
```
[the remark this answers is not in the document]
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Actually it is a good practice and code quality thing to add brackets.

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/utilities/GitLabUtils/index.js [hunk 13]
```diff
@@ -0,0 +1,364 @@
+
+// Import JSON structure
+import { structure, organizations, members, contribution, repos, calendarEntry, yearEntry} from "./Objects";
+let data = structure;
+
+//> Parser functions
+// Parse Json to DOM Object
+const parseJsonToDOM = (json) => {
```
[the remark this answers is not in the document]
Comment (Aichnerc) [reply src/utilities/GitLabUtils/index.js]: Actually it is a good practice and code quality thing to add brackets.

```
// Bad 
a => {
}
// Good 
(a) => {
}
```

Ref: https://eslint.org/docs/rules/arrow-parens

File: src/App.js [hunk 14]
```diff
@@ -1,8 +1,8 @@
 //> React
 // Contains all the functionality necessary to define React components
-import React from 'react';
+import React from "react";
```
Comment (Aichnerc) [inline src/App.js]: Please use single quotes `'` instead of double quotes `"`

File: src/App.js [hunk 15]
```diff
@@ -1,8 +1,8 @@
...
 // DOM bindings for React Router
-import { BrowserRouter as Router } from 'react-router-dom';
+import { BrowserRouter as Router } from "react-router-dom";
```
Comment (Aichnerc) [inline src/App.js]: Please use single quotes `'` instead of double quotes `"`

File: src/App.js [hunk 16]
```diff
@@ -15,28 +15,28 @@ import {
   Navbar,
 } from './components/molecules';
 */
-// Routes
-import Routes from './Routes';
 
+//> Routes
+import Routes from "./Routes";
```
Comment (Aichnerc) [inline src/App.js]: Please use single quotes `'` instead of double quotes `"`

File: src/App.js [hunk 17]
```diff
@@ -15,28 +15,28 @@ import {
...
+
+//> Tests
+// Testing retrieving activity data from gitLab
+import "./App.test";
```
Comment (Aichnerc) [inline src/App.js]: Please use single quotes `'` instead of double quotes `"`

File: src/App.test.js [hunk 18]
```diff
@@ -0,0 +1,11 @@
+import * as gitLab from "./utilities/GitLabUtils";
```
Comment (Aichnerc) [inline src/App.test.js]: Please use single quotes `'` instead of double quotes `"`

File: src/utilities/GitLabUtils/index.js [hunk 19]
```diff
@@ -0,0 +1,356 @@
+// Import JSON structure
+import {
+  structure,
+  organizations,
+  members,
+  contribution,
+  repos,
+  calendarEntry,
+  yearEntry
+} from "./Objects";
```
Comment (Aichnerc) [inline src/utilities/GitLabUtils/index.js]: Please use single quotes `'` instead of double quotes `"`

**Pull request merged**